enhancements/microshift/feature-csi-snapshots.md

```diff
@@ -0,0 +1,285 @@
+
+The VolumeSnapshot CRD is namespaced. A developer uses the CRD as a distinct request for a snapshot.
+
+```yaml
+apiVersion: snapshot.storage.k8s.io/v1
+kind: VolumeSnapshot
+metadata:
+  name: mysnap
+spec:
+  volumeSnapshotClassName: csi-hostpath-snap [1] 
+  source:
+    persistentVolumeClaimName: myclaim [2]
+```
+
+1. The request for a particular class by the volume snapshot. If the **volumeSnapshotClassName** 
+setting is absent and there is a default volume snapshot class, a snapshot is created with the
+default volume snapshot class name. But if the field is absent and no default volume snapshot class
+exists, then no snapshot is created.
+2. The name of the **PersistentVolumeClaim** object bound to a persistent volume. This defines what
+you want to create a snapshot of. Required for dynamically provisioning a snapshot.
+
+#### VolumeSnapshotContent
+
+A snapshot taken of a volume in the cluster that has been provisioned by a cluster administrator.
+
+Similar to the PersistentVolume object, the VolumeSnapshotContent CRD is a cluster resource that
+points to a real snapshot in the storage back end.
+
+For manually pre-provisioned snapshots, a cluster administrator creates a number of
+VolumeSnapshotContent CRDs. These carry the details of the real volume snapshot in the storage system.
+
+The VolumeSnapshotContent CRD is not namespaced and is for use by a cluster administrator.
+
+```yaml
+apiVersion: snapshot.storage.k8s.io/v1
+kind: VolumeSnapshotContent
+metadata:
+  name: new-snapshot-content-test
+  annotations:
+    - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
+spec:
+  deletionPolicy: Delete
+  driver: hostpath.csi.k8s.io
+  source:
+    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
+  sourceVolumeMode: Filesystem
+  volumeSnapshotRef:
+    name: new-snapshot-test
+    namespace: default
+```
+
+### Risks and Mitigations
+
+- Increased overhead.  The CSI snapshot controller and sidecar increase the on-disk footprint by a
+total of 360Mb.  At idle, the components use a negligible amount of CPU (>1% CPU on a 2 core machine)
+and roughly 18Mb.  Under load, the 
+
+### Drawbacks
+
+## Design Details
+
+### Test Plan
+
+Snapshotting functionality will be tested under the Openshift conformance test suite.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Full GA
+
+- Available by default
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+The CSI Snapshot controller and sidecar are pieces of the Kubernetes CSI implementation.  OpenShift
+maintains downstream versions of these images and tracks them as part of OCP releases.
+
+### Version Skew Strategy
+
+The CSI Snapshot version is tracked by the ocp-release image.  This provides sufficient guardrails
+against version skew.
```

Comment:
The CRDs are versioned with Kube.

Reply:
Does that mean we have to worry about the kube rebase being applied to cluster-csi-snapshot-controller-operator before MicroShift can take the same update?

Reply:
I think I'm not communicating my concern clearly.

I can't tell which repository the CRDs for these new types are going to come from. I'm worried that if they are not in the `openshift/kubernetes` repository then when OpenShift updates its kube version we will have some version skew between components that we are vendoring. We already have a similar situation with the route-controller-manager that means MicroShift cannot update its kube version until the RCM does. Is this a new case that will require updating something else, like a storage-related operator?

Reply:
Is this only in the vendoring context? The design here would be to deploy the csi-controller as a pod on the cluster, not to embed it.  We wouldn't be rebasing code, just updating the image digests.

I see the CRDs are also defined in [openshift/kubernetes](https://github.com/openshift/kubernetes/tree/15f19ea2dd700767e5337502aec753d2a6e26905/cluster/addons/volumesnapshots/crds), so we could pull them from here instead to stay on the safe side.

Reply:
Regardless of how we run the component, we need to bring the manifests into the MicroShift repository and we want them to match the code that is being run.

Reply:
Yeah, understood.  We should stick with sourcing them from the openshift-csi-controller-operator in that case.  Reason being that the openshift/kubernetes CRDs are versioned with the upstream container image [here](https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/cluster/addons/volumesnapshots/volume-snapshot-controller/volume-snapshot-controller-deployment.yaml#LL25C9-L25C9), which we don't want to use since Openshift already maintains a downstream.  The controller-operator is versioned with the Openshift versions of the images from quay.io/openshift4/, [here](https://github.com/openshift/cluster-csi-snapshot-controller-operator/blob/eaaf7260464bd6cb6126dfd7259cb6217eb25340/manifests/07_deployment.yaml#L42-L49).

Reply:
Yes, that's what I was thinking.  Wherever we get the CRD manifests from, we should also use the relative image, as they are versioned together.  That way, we can be pretty certain that the CRDs and the container versions are synced.